synapse/storage/databases/main/devices.py

```diff
@@ -90,20 +92,44 @@ def __init__(
 
         # In the worker store this is an ID tracker which we overwrite in the non-worker
         # class below that is used on the main process.
-        self._device_list_id_gen = StreamIdGenerator(
-            db_conn,
-            hs.get_replication_notifier(),
-            "device_lists_stream",
-            "stream_id",
-            extra_tables=[
-                ("user_signature_stream", "stream_id"),
-                ("device_lists_outbound_pokes", "stream_id"),
-                ("device_lists_changes_in_room", "stream_id"),
-                ("device_lists_remote_pending", "stream_id"),
-                ("device_lists_changes_converted_stream_position", "stream_id"),
-            ],
-            is_writer=hs.config.worker.worker_app is None,
-        )
+        self._device_list_id_gen: AbstractStreamIdGenerator
+        if isinstance(database.engine, PostgresEngine):
```

Comment:
Following the pattern I noticed in other files where `MultiWriterIdGenerator` is used, although I think we could refactor this and create a helper function to get the right `AbstractStreamIdGenerator` depending on the engine, similarly to what `build_sequence_generator` is doing. 
It would be less verbose and abstract some complexity away from the callee by making `MultiWriterIdGenerator` and `StreamIdGenerator` implementation details.
If you agree with this I could start by making the refactor on develop on then use the new pattern here.

Note that I think that eventually, `StreamIdGenerator` and `MultiWriterIdGenerator` should become private to prevent someone from using `StreamIdGenerator` and breaking the multi node setup.